gui/src/app/CompileContext/CompileContextProvider.tsx

```diff
@@ -0,0 +1,86 @@
+import { ProjectContext } from "@SpCore/ProjectContextProvider";
+import compileStanProgram from "@SpStanc/compileStanProgram";
+import {
+  FunctionComponent,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { CompileContext, CompileStatus } from "./CompileContext";
+
+type CompileContextProviderProps = {
+  // none
+};
+
+export const CompileContextProvider: FunctionComponent<
+  PropsWithChildren<CompileContextProviderProps>
+> = ({ children }) => {
+  const { data } = useContext(ProjectContext);
+  const [compileStatus, setCompileStatus] = useState<CompileStatus>("");
+  const [
+    theStanFileContentThasHasBeenCompiled,
+    setTheStanFileContentThasHasBeenCompiled,
+  ] = useState<string>("");
+  const [compileMessage, setCompileMessage] = useState<string>("");
+  const [compiledMainJsUrl, setCompiledMainJsUrl] = useState<
+    string | undefined
+  >(undefined);
+  const [validSyntax, setValidSyntax] = useState<boolean>(false);
+
+  useEffect(() => {
+    // if the compiled content is not the same as the current content,
+    // then the state should not be compiled or failed
+    if (data.stanFileContent !== theStanFileContentThasHasBeenCompiled) {
+      if (compileStatus === "compiled" || compileStatus === "failed") {
+        setCompileStatus("");
+        setCompiledMainJsUrl("");
+      }
+    }
+  }, [
+    data.stanFileContent,
+    theStanFileContentThasHasBeenCompiled,
+    compileStatus,
+    setCompiledMainJsUrl,
+  ]);
+
+  const handleCompile = useCallback(async () => {
+    setCompileStatus("compiling");
+    await new Promise((resolve) => setTimeout(resolve, 500)); // for effect
+    const onStatus = (msg: string) => {
+      setCompileMessage(msg);
+    };
+    const stanWasmServerUrl =
+      localStorage.getItem("stanWasmServerUrl") ||
+      "https://trom-stan-wasm-server.magland.org";
```

Comment:
This is relying on the fact that the compilation control sets local storage, but should this setting just live in this context rather than in that UI?

Reply:
With https://github.com/flatironinstitute/stan-playground/pull/191/commits/675b37f07b9d7fe344b20796e4dba9dc8c7deb70 I moved stanWasmServerUrl into CompileContext and it simplified a lot, and the interaction with stanWasmServerUrl in localStorage is now in one place. In that commit I also fixed some issues with the compile server config dialog - it seems to be working smoothly now.

Reply:
I realized that some of those things had already been fixed in main branch and there were some conflicts. Resolved the conflicts.